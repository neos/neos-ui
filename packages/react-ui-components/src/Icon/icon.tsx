import React, {PureComponent} from 'react';

import {FontAwesomeIconProps} from '@fortawesome/react-fontawesome';
import {PickDefaultProps, Omit} from '../../types';
import FontAwesomeIcon from './fontAwesomeIcon';
import ResourceIcon, {ResourceIconProps} from './resourceIcon';

type IconSize = 'xs' | 'sm' | 'lg' | '2x' | '3x';
type IconPadding = 'none' | 'left' | 'right';
type IconColor = 'default' | 'warn' | 'error' | 'primaryBlue';

export interface IconTheme {
    readonly icon: string;
    readonly 'icon--huge': string;
    readonly 'icon--large': string;
    readonly 'icon--big': string;
    readonly 'icon--small': string;
    readonly 'icon--tiny': string;
    readonly 'icon--paddedLeft': string;
    readonly 'icon--paddedRight': string;
    readonly 'icon--spin': string;
    readonly 'icon--color-warn': string;
    readonly 'icon--color-error': string;
    readonly 'icon--color-primaryBlue': string;
}

export interface IconProps extends Omit<FontAwesomeIconProps, 'icon'> {
    /**
     * We use the react component FortAwesome provides to render icons.
     * we will pass down all props to the component via {...rest} to expose it's api
     * https://github.com/FortAwesome/react-fontawesome/blob/master/README.md
     */

    /**
     * The identifier of the icon to render.
     * Can be a font-awesome icon identifier or a asset resource.
     */
    readonly icon?: string;

    /**
     * The (accessibility) label for this icon
     */
    readonly label?: string;

    /**
     * Controls the padding around the icon in a standardized way.
     */
    readonly padded?: IconPadding;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     *  Adjust the color of the icon
     */
    readonly color?: IconColor;

    /**
     *  Adjust the size of the icon
     */
    readonly size?: IconSize;

    /**
     *  An optional css theme to be injected.
     */
    readonly theme?: IconTheme;
}

type DefaultProps = PickDefaultProps<IconProps, 'color' | 'padded' | 'size'>;

export const defaultProps: DefaultProps = {
    color: 'default',
    padded: 'none',
    size: 'sm'
};

class Icon extends PureComponent<IconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {icon} = this.props;

        if (icon && icon.substr(0, 11) === 'resource://') {
            return <ResourceIcon {...this.props as ResourceIconProps} />;
        } else {
            return <FontAwesomeIcon {...this.props} />;
        }
    }
}

export default Icon;
