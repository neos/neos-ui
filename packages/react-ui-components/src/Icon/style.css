.icon {
    composes: reset from './../reset.css';
    display: inline-block;
    font: normal normal normal FontAwesome;
    font-size: 14px/1;
    text-rendering: auto;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
}
.icon--paddedLeft {
    margin-left: .75em;
}
.icon--paddedRight {
    margin-right: .75em;
}

.icon--color-warn {
    color: var(--colors-Warn);
}

.icon--color-error {
    color: var(--colors-Error);
}

.icon--color-primaryBlue {
    color: var(--colors-PrimaryBlue);
}

.icon--huge {
    svg {
        height: 3em;
    }
}

.icon--large {
    svg {
        height: 2em;
    }
}

.icon--big {
    svg {
        height: 1.33em;
    }
}

.icon--small {
    svg {
        height: 1em;
        max-height: 1em;
        max-width: 100%;
    }
}

.icon--tiny {
    svg {
        height: .75em;
    }
}

.icon--resource {
    display: inline-grid;
    width: 100%;
    justify-content: center;
}
