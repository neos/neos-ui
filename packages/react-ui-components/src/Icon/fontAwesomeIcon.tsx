import {config, IconName, IconPrefix, library} from '@fortawesome/fontawesome-svg-core';
import '@fortawesome/fontawesome-svg-core/styles.css';
import {fab} from '@fortawesome/free-brands-svg-icons';
import {far} from '@fortawesome/free-regular-svg-icons';
import {fas} from '@fortawesome/free-solid-svg-icons';

import {FontAwesomeIcon as FontAwesomeIconOriginComponent} from '@fortawesome/react-fontawesome';
import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';

import {IconProps, defaultProps} from './icon';
import mapper from './mapper';

// tslint:disable:no-object-mutation
config.autoAddCss = false;
config.familyPrefix = 'neos-fa' as IconPrefix;
config.replacementClass = 'neos-svg-inline--fa';
// tslint:enable:no-object-mutation

library.add(fab, fas, far);

class FontAwesomeIcon extends PureComponent<IconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {padded, theme, label, icon, className, color, ...rest} = this.props;
        const iconClassName = icon;
        const classNames = mergeClassNames(
            theme!.icon,
            iconClassName,
            className,
            {
                [theme!['icon--paddedLeft']]: padded === 'left',
                [theme!['icon--paddedRight']]: padded === 'right',
                [theme!['icon--color-warn']]: color === 'warn',
                [theme!['icon--color-error']]: color === 'error',
                [theme!['icon--color-primaryBlue']]: color === 'primaryBlue',
            }
        );

        return <FontAwesomeIconOriginComponent icon={icon ? this.getIconProp(icon) as any : 'question'} aria-label={label} className={classNames} {...rest} />;
    }

    private readonly getIconProp = (icon: string): any => {
        const mappedIcon = mapper(icon);
        const iconArray = mappedIcon.split(' ');
        if (iconArray.length > 1) {
            const prefix = iconArray[0];
            const processedIcon = iconArray[1].startsWith('fa-') ? iconArray[1].substr(3) : iconArray[1];
            return [prefix as IconPrefix, processedIcon as IconName];
        } else {
            const prefix: IconPrefix = 'fas';
            return [prefix, mappedIcon as IconName];
        }
    }
}

export default FontAwesomeIcon;
