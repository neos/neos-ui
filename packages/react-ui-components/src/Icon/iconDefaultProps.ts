import {IconProps} from './icon';
import {PickDefaultProps} from '../../types';

type DefaultProps = PickDefaultProps<IconProps, 'color' | 'padded' | 'size'>;

// extracted because of resulting in a circular dependency in jest
// internal - use `import {defaultProps} from "./icon"` instead
export const defaultProps: DefaultProps = {
    color: 'default',
    padded: 'none',
    size: 'sm'
};
