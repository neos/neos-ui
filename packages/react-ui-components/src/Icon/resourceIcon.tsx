import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';
import { ReactSVG } from 'react-svg';

import {IconProps, IconTheme, defaultProps} from './icon';
import {Omit} from '../../types';

interface ResourceIconTheme extends IconTheme {
    readonly 'icon--resource': string;
}

export interface ResourceIconProps extends Omit<IconProps, 'theme'> {
    readonly theme?: ResourceIconTheme;
}

class ResourceIcon extends PureComponent<ResourceIconProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element |Â null {
        const {padded, theme, label, icon, className, color, size} = this.props;

        if (!icon || icon.substr(0, 11) !== 'resource://') {
            return null;
        }

        const iconResourcePath = '/_Resources/Static/Packages/' + icon.substr(11);
        const classNames = mergeClassNames(
            theme!.icon,
            className,
            {
                [theme!['icon--resource']]: icon,
                [theme!['icon--paddedLeft']]: padded === 'left',
                [theme!['icon--paddedRight']]: padded === 'right',
                [theme!['icon--color-warn']]: color === 'warn',
                [theme!['icon--color-error']]: color === 'error',
                [theme!['icon--color-primaryBlue']]: color === 'primaryBlue',
                [theme!['icon--huge']]: size === '3x',
                [theme!['icon--large']]: size === '2x',
                [theme!['icon--big']]: size === 'lg',
                [theme!['icon--small']]: size === 'sm',
                [theme!['icon--tiny']]: size === 'xs'
            }
        );

        return <ReactSVG src={iconResourcePath} aria-label={label} className={classNames} wrapper="span" />;
    }
}

export default ResourceIcon;
