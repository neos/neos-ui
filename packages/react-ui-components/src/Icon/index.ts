import {themr} from '@friendsofreactjs/react-css-themr';

import identifiers from '../identifiers';
import Icon from './icon';

import style from './style.css';

export default themr(identifiers.icon, style)(Icon);
