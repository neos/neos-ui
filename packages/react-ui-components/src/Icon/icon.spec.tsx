import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import Icon, {defaultProps, IconProps} from './icon';

describe('<Icon/>', () => {
    const props: IconProps = {
        ...defaultProps,
        icon: 'fooIconClassName',
        theme: {
            'icon': 'iconClassName',
            'icon--big': 'bigClassName',
            'icon--color-error': 'color-errorClassName',
            'icon--color-primaryBlue': 'color-primaryBlueClassName',
            'icon--color-warn': 'color-warnClassName',
            'icon--paddedLeft': 'paddedLeftClassName',
            'icon--paddedRight': 'paddedRightClassName',
            'icon--small': 'smallClassName',
            'icon--spin': 'spinClassName',
            'icon--tiny': 'tinyClassName',
            'icon--large': 'largeClassName',
            'icon--huge': 'hugeClassName'
        }
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<Icon {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<Icon {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });
});

