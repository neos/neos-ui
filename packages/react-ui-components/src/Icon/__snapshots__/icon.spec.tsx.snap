// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon/> should render correctly. 1`] = `
<FontAwesomeIcon
  color="default"
  icon="fooIconClassName"
  padded="none"
  size="sm"
  theme={
    Object {
      "icon": "iconClassName",
      "icon--big": "bigClassName",
      "icon--color-error": "color-errorClassName",
      "icon--color-primaryBlue": "color-primaryBlueClassName",
      "icon--color-warn": "color-warnClassName",
      "icon--huge": "hugeClassName",
      "icon--large": "largeClassName",
      "icon--paddedLeft": "paddedLeftClassName",
      "icon--paddedRight": "paddedRightClassName",
      "icon--small": "smallClassName",
      "icon--spin": "spinClassName",
      "icon--tiny": "tinyClassName",
    }
  }
/>
`;
